### PR TITLE
Loading data from Big Data Viewer

### PR DESCRIPTION
**Pull request by IgorTatarnikov**
Before submitting a pull request (PR), please read the [contributing guide](https://github.com/brainglobe/.github/blob/main/CONTRIBUTING.md).

Please fill out as much of this template as you can, but if you have any problems or questions, just leave a comment and we will help out :)

## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
This PR allows loading data from h5 files produced by mesoSPIM.

**What does this PR do?**
Create a tile class:
- This class holds the resolution pyramid and metadata pertaining to each tile, such as the position within the stitched image.
- The initial position is set based on the stage coordinates provided by the mesoSPIM in the .h5_meta.txt file.

Create an ImageMosaic class:
- This class will hold most of the logic for the package.
- It will hold a list of Tile objects as well as paths to various files/file objects required for downstream operations.

Create a StitchingWidget class that serves as a napari plugin:
- Downsampled data is passed to be displayed in `napari`.
- The widget keeps a list of tile layers that it's working with to allow updating positions. 

## How has this PR been tested?

Tests have been written that try to cover most basic cases for the added code. This package will use test data in the form of a downsampled tiled acquisition of one brain. The test data is stored on [GIN](https://gin.g-node.org/IgorTatarnikov/brainglobe-stitch-test/raw/master/brainglobe-stitch/brainglobe-stitch-test-data.zip).

A full size example can be found on ceph at `/ceph/neuroinformatics/neuroinformatics/igort/stitching_data/Brain2`

## Is this a breaking change?

No.

## Does this PR require an update to the documentation?

Yes, docs have not been written at this time.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality (unit & integration)
- [ ] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

**Review by alessandrofelder — APPROVED** [review]
AFAICT this looks good.
Think all my minor comments are requests to document slightly more explicitly, to help code legibility when it's not as present in our heads :)

File: brainglobe_stitch/file_utils.py [hunk 1]
```diff
@@ -8,70 +13,109 @@
...
+                grp.require_dataset(
+                    f"{i}/cells",
+                    dtype=prev_resolution.dtype,
+                    shape=(
+                        (prev_resolution.shape[0] + 1)
+                        // downsampling_factors[2],
+                        (prev_resolution.shape[1] + 1)
+                        // downsampling_factors[1],
+                        (prev_resolution.shape[2] + 1)
+                        // downsampling_factors[0],
```
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply brainglobe_stitch/file_utils.py:82]: It's related to how the data is stored by the mesoSPIM. The dataset is stored in z, y, x order while the associated resolution and subdivision data is stored in x, y, z order. I tried to add a note.

File: pyproject.toml [hunk 2]
```diff
@@ -21,6 +21,19 @@ classifiers = [
     "Topic :: Scientific/Engineering :: Image Processing",
 ]
 
+dependencies = [
+    "napari>=0.4.18",
+    "brainglobe-utils>=0.3.4",
+    "h5py",
+    "superqt",
```
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply pyproject.toml]: I started using `superqt` to add some features later, but I'll add it as a requirement then!

File: brainglobe_stitch/file_utils.py [hunk 3]
```diff
@@ -8,70 +13,109 @@
...
-
-
-def parse_mesospim_metadata(meta_file_name: Path):
+def create_pyramid_bdv_h5(
+    input_file: Path,
+    yield_progress: bool = False,
+):
+    """
+    Create a resolution pyramid for a Big Data Viewer HDF5 file.
+
```
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply brainglobe_stitch/file_utils.py:34]: Done

File: brainglobe_stitch/file_utils.py [hunk 4]
```diff
@@ -8,70 +13,109 @@
...
+    yield_progress: bool, optional
+        Whether to yield progress.
+    """
+    resolutions_array = np.array(
+        [[1, 1, 1], [2, 2, 2], [4, 4, 4], [8, 8, 8], [16, 16, 16]]
+    )
+
+    subdivisions_array = np.array(
+        [[32, 32, 16], [32, 32, 16], [32, 32, 16], [32, 32, 16], [32, 32, 16]]
+    )
```
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply brainglobe_stitch/file_utils.py]: I've changed both to be a function parameter with a default. I based the defaults on how `BigStitcher` and `BigDataViewer` save the h5 files.

File: tests/test_unit/conftest.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+TEMP_DIR = Path.home() / "temp_test_directory"
+TEST_DATA_URL = "https://gin.g-node.org/IgorTatarnikov/brainglobe-stitch-test/raw/master/brainglobe-stitch/brainglobe-stitch-test-data.zip"
+
+
+@pytest.fixture(scope="session", autouse=True)
+def download_test_data():
+    TEMP_DIR.mkdir(exist_ok=True)
+    pooch.retrieve(
+        TEST_DATA_URL,
+        known_hash=None,
```
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply tests/test_unit/conftest.py]: Done

File: brainglobe_stitch/stitching_widget.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+            blending="translucent",
+            contrast_limits=[0, 4000],
+            multiscale=False,
+        )
+        tile_layer.translate = tile_position
+
+        yield tile_layer
+
+
+class StitchingWidget(QWidget):
```
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply brainglobe_stitch/stitching_widget.py:58]: Done!

File: brainglobe_stitch/stitching_widget.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+)
+
+from brainglobe_stitch.file_utils import (
+    check_mesospim_directory,
+    create_pyramid_bdv_h5,
+)
+from brainglobe_stitch.image_mosaic import ImageMosaic
+
+
+def add_tiles_from_mosaic(
```
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply brainglobe_stitch/stitching_widget.py:30]: I was getting some weird errors when trying to run this in a worker thread until I pulled it out of the class itself :shrug:

File: tests/test_unit/test_image_mosaic.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+import pytest
+
+from brainglobe_stitch.image_mosaic import ImageMosaic
+
+NUM_TILES = 8
+NUM_RESOLUTIONS = 5
+NUM_CHANNELS = 2
+TILE_SIZE = (107, 128, 128)
+EXPECTED_TILE_CONFIG = [
```
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply tests/test_unit/test_image_mosaic.py:11]: I've added some comments describing the test file and the expected tile position. The tiles overlap by 10% (13 pixels) in x, y, and there's no tiling in z, so naively they all end up starting at `z=0`. `EXPECTED_TILE_CONFIG` is checking the output of a file that's ingested by `BigStitcher` so it needs the tile positions to be in `x, y, z` order while the `EXPECTED_TILE_POSITIONS` are for `napari` so they're all in `z, y, x` order.

**Pull request merged**